### PR TITLE
Adding NATS Microservices

### PR DESCRIPTION
**Pull request by EmilJimenez21**
I've gone ahead and implemented NATS Micro inspired by [NATS.go](https://github.com/nats-io/nats.go/)

[+] Updated README.md to include an example of a microservice implementation. 
[+] Updated Client.php to include the service() function 
[+] Implemented the required standards outlined by NATS Micro to support all "nats micro" cli calls.

### CODE REVIEW

**Comment by nekufa** [comment]
@EmilJimenez21 thanks for the contribution!
code looks good for me, but it's not tested
would you like to add some tests?

**Comment by nekufa** [comment]
@EmilJimenez21 i'm not familar with microservice approach, but it looks like that if you can describe microservice, somebody can call it. i think that infinitive loop can be skipped for test purposes, but when you call microservice endpoint and you can call client process with timeout option, so client will process some requests and then stop. 

like we test for basic features like publish/subscribe
https://github.com/basis-company/nats.php/blob/main/tests/Functional/SubjectTest.php#L64-L79

File: src/Client.php [hunk 1]
```diff
@@ -260,4 +263,11 @@ public function disconnect(): self
 
         return $this;
     }
+
+    public function service(string $name, string $description, string $version): Service
+    {
+        $this->services[$name] = new Service($this, $name, $description, $version);
```
[the remark this answers is not in the document]
Comment (EmilJimenez21) [reply src/Client.php]: Fixed the instances where it could occur

**Comment by EmilJimenez21** [comment]
@nekufa I wrote some tests. Let me know your thoughts

**Comment by nekufa** [comment]
Hi @EmilJimenez21 sorry for the delay, i'm still out of the office, will get back soon.
Can this one wait several days or you need to merge this asap?

**Comment by EmilJimenez21** [comment]
@nekufa Take your time, I'm in no rush. Enjoy the time off!

File: src/Service/Service.php [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    }
+
+    private function stats(): array
+    {
+        return [
+            'type' => 'io.nats.micro.v1.stats_response',
+            'name' => $this->name,
+            'id' => $this->id,
+            'version' => $this->version,
+            'endpoints' => array_reduce($this->endpoints, function ($carry, ServiceEndpoint $endpoint) {
```
Comment (nekufa) [inline src/Service/Service.php:96]: maybe simplify this line with introduce endpoints var and fill it using foreach?

File: src/Service/Service.php [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        if ($id == '') {
+            return "\$SRV.$verb.$name";
+        }
+
+        return "\$SRV.$verb.$name.$id";
+    }
+
+    public function run() : void
+    {
+        print_r("$this->name is ready to accept connections\n");
```
Comment (nekufa) [inline src/Service/Service.php:218]: standard way to add debug information should be optional client.logger usage when present.
see `Connection.php` source for an example

File: src/Service/ServiceEndpoint.php [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    private float $processing_time = 0;
+
+    private Client|Queue $subscription;
+
+    public function __construct(
+        private readonly Service $service,
+        private readonly string  $name,
+        private readonly string  $subject,
+        private                  $endpointHandler,
+        private readonly string  $queue_group = 'q'
```
Comment (nekufa) [inline src/Service/ServiceEndpoint.php:26]: this library use another code style, see phpcs.xml and validate it using phpcs binary.
indentation using spaces should be removed

File: src/Service/ServiceEndpoint.php [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+                $start = microtime(true);
+
+                // Update the endpoint metrics
+                $this->num_requests += 1;
+
+                // Setup the response
+                $response = "";
+
+                switch ($this->endpointHandler) {
+                    case is_string($this->endpointHandler):
```
Comment (nekufa) [inline src/Service/ServiceEndpoint.php:43]: let's replace this case with `is_subclass_of` check and add default case with logic exception

File: tests/Functional/ServiceTest.php [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+class ServiceTest extends FunctionalTestCase
+{
+    private bool $tested = false;
+
+    private function createTestService(): Service
+    {
+        /** @var Client $client */
+        $client = $this->createClient([
+            'host' => 'hermes.internal'
```
Comment (nekufa) [inline tests/Functional/ServiceTest.php:19]: you can configuration local runner using phpunit.xml via `NATS_HOST`\
just copy phpunit.xml.dist and replace all you need

File: tests/Functional/ServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+            );
+
+        $service->client->publish('v1.test', '');
+
+        $response = $service->client->process(1);
+
+        $this->assertTrue($response['success']);
+    }
+
+    public function testServiceRequestReplyClass()
```
Comment (nekufa) [inline tests/Functional/ServiceTest.php:71]: this test is for instance, not class. i would like to fix method name)

**Comment by EmilJimenez21** [comment]
@nekufa Thank you for all the feedback! I implemented the changes you highlighted

**Comment by nekufa** [comment]
@EmilJimenez21 thanks!\
just tested localy and find out some issues with same name for microservice.\
i'm fixing it now and release is on the way

**Pull request merged**